File: scb-web/src/utils/building.js
export function generateFloors(start, end) {
  const floors = []
  for (let i = start; i <= end; i++) {
    if (i !== 0) floors.push(i)
  }
  return floors
}

// exceptions: { 21: 2, 1: 3, ... } (층 번호 → 세대 수)
export function createBuilding({ name, startFloor, endFloor, defaultUnits, exceptions = {}, pinanFloors = [] }) {
  const units = []
  const floors = generateFloors(startFloor, endFloor)
  let maxUnits = defaultUnits

  for (const floor of floors) {
    const count = exceptions[floor] !== undefined ? Number(exceptions[floor]) : defaultUnits
    if (count > maxUnits) maxUnits = count
    for (let i = 1; i <= count; i++) {
      units.push({
        id: `${floor}-${i}`,
        floor,
        unitIndex: i,
        number: floor * 100 + i, // 1층1호=101, 21층1호=2101
        status: 'unknown',
        memo: '',
        memoPinned: false,
        vulnerable: false,
      })
    }
  }

  return { name, startFloor, endFloor, defaultUnits, exceptions, units, maxUnits, pinanFloors }
}

export function getFloorsDesc(b) {
  return generateFloors(b.startFloor, b.endFloor).sort((a, z) => z - a)
}

export function getUnitsOnFloor(units, floor) {
  return units.filter(u => u.floor === floor).sort((a, b) => a.unitIndex - b.unitIndex)
}

File: safe_check_board/lib/models/building.dart
class Building {
  final String name;
  final int startFloor;
  final int endFloor;
  final int defaultUnits;  // 층(구역)당 세대(구획)수
  final bool isHorizontal; // 가로형(시장·의료) vs 세로형(아파트·빌딩)

  Building({
    required this.name,
    required this.startFloor,
    required this.endFloor,
    required this.defaultUnits,
    List<int>? pinanFloors, // 하위호환용 (무시)
    this.isHorizontal = false,
  });

  /// 옥상층 번호 (세로형만, endFloor + 1)
  int? get rooftopFloor => isHorizontal ? null : endFloor + 1;

  /// 유효 층 목록 (0층 제외, 세로형은 옥상층 포함)
  List<int> get floors {
    final result = <int>[];
    for (int i = startFloor; i <= endFloor; i++) {
      if (i != 0) result.add(i);
    }
    if (!isHorizontal) result.add(endFloor + 1); // 옥상층
    return result;
  }

  /// 위→아래 순서 (내림차순)
  List<int> get floorsDescending {
    final result = floors;
    result.sort((a, b) => b.compareTo(a));
    return result;
  }

}
